tests/tso/client_test.go

```diff
@@ -0,0 +1,173 @@
+// Copyright 2023 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tso
+
+import (
+	"context"
+	"math"
+	"strings"
+	"sync"
+	"testing"
+
+	"github.com/pingcap/failpoint"
+	"github.com/stretchr/testify/suite"
+	pd "github.com/tikv/pd/client"
+	"github.com/tikv/pd/client/testutil"
+	tso "github.com/tikv/pd/pkg/mcs/tso/server"
+	"github.com/tikv/pd/pkg/utils/tsoutil"
+	"github.com/tikv/pd/tests"
+	"github.com/tikv/pd/tests/mcs"
+)
+
+type tsoClientTestSuite struct {
+	suite.Suite
+	legacy bool
+
+	ctx    context.Context
+	cancel context.CancelFunc
+	// The PD cluster.
+	cluster *tests.TestCluster
+	// The TSO service in microservice mode.
+	tsoServer        *tso.Server
+	tsoServerCleanup func()
+
+	client pd.TSOClient
+}
+
+func TestLegacyTSOClient(t *testing.T) {
+	suite.Run(t, &tsoClientTestSuite{
+		legacy: true,
+	})
+}
+
+func TestMicroserviceTSOClient(t *testing.T) {
+	suite.Run(t, &tsoClientTestSuite{
+		legacy: false,
+	})
+}
+
+func (suite *tsoClientTestSuite) SetupSuite() {
+	re := suite.Require()
+
+	var err error
+	suite.ctx, suite.cancel = context.WithCancel(context.Background())
+	switch suite.legacy {
```

Comment:
interesting. We use switch for bool. I'm ok with it which seems to be personal preference.